### PR TITLE
chore(code coverage): enable coverage remapping to source via sourcemap

### PR DESCRIPTION
**Pull request by kwonoj**
- update script enables test coverage remapping to original (es6) source
- minor update to document, code

This PR introduces [`remap-istanbul` module](https://github.com/SitePen/remap-istanbul) allows to remap test coverage into original source code instead of transpiled results. (`istanbul` side discussion is ongoing at https://github.com/gotwarlost/istanbul/issues/212) Due to current build processes, this PR does not resolves remapping issue fully though.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,19 +6,19 @@
   "scripts": {
     "build_all": "npm run build_es6 && npm run build_amd && npm run build_cjs && npm run build_global",
     "build_amd": "rm -rf dist/amd && tsc typings/es6-shim/es6-shim.d.ts src/Rx.ts -m amd --outDir dist/amd --sourcemap --target ES5",
-    "build_cjs": "rm -rf dist/cjs && babel dist/es6 --out-dir dist/cjs --modules common --sourceMaps --loose all && node lib/copy_dts.js",
-    "build_es6": "rm -rf dist/es6 && tsc src/Rx.ts src/Rx.KitchenSink.ts --outDir dist/es6 --target ES6 -d",
+    "build_cjs": "rm -rf dist/cjs && babel dist/es6 --out-dir dist/cjs --modules common --source-maps --loose all && node lib/copy_dts.js",
+    "build_es6": "rm -rf dist/es6 && tsc src/Rx.ts src/Rx.KitchenSink.ts --outDir dist/es6 --sourceMap --target ES6 -d",
```
Comment (kwonoj) [inline package.json]: Found cjs / es6 modules didn't have sourcemap until now..

File: package.json [hunk 2]
```diff
@@ -6,19 +6,19 @@
...
     "build_test": "rm -rf dist/ && npm run lint && npm run build_es6 && npm run build_cjs && jasmine",
-    "build_cover": "rm -rf dist/ && npm run lint && npm run build_es6 && npm run build_cjs && istanbul cover jasmine",
+    "build_cover": "rm -rf dist/ && npm run lint && npm run build_es6 && npm run build_cjs && npm run cover",
     "build_docs": "./docgen.sh",
     "lint_perf": "eslint perf/",
     "lint_spec": "eslint spec/**/*.js",
     "lint_src": "tslint -c .tslintrc src/**/*.ts",
     "lint": "npm run lint_src && npm run lint_spec && npm run lint_perf",
-    "cover": "istanbul cover jasmine",
+    "cover": "istanbul cover -x \"*-spec.js index.js *-helper.js\" jasmine",
```
Comment (kwonoj) [inline package.json]: Run test coverage exclude actual test cases.

File: src/testing/HotObservable.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import Subject from '../Subject';
-import Subscriber from './Subscriber';
+import Subscriber from '../Subscriber';
```
Comment (kwonoj) [inline src/testing/HotObservable.ts]: Typo correction - curious how tsc build was passed while module import path resolving fails.

File: .travis.yml [hunk 4]
```diff
@@ -20,4 +20,5 @@ script:
   - npm run build_cover
 
 after_script:
-  - cat ./coverage/lcov.info | ./node_modules/coveralls/bin/coveralls.js
+  - ./node_modules/.bin/remap-istanbul -i coverage/coverage.json -o coverage/coverage-remapped.json && ./node_modules/.bin/remap-istanbul -i coverage/coverage.json -o coverage/coverage-remapped.lcov -t lcovonly && ./node_modules/.bin/remap-istanbul -i coverage/coverage.json -o coverage/coverage-remapped -t html
```
Comment (kwonoj) [inline .travis.yml]: Instead of send coverage report directly, now it remaps with sourcemap first.

**Comment by kwonoj** [comment]
It is expected to see coverage number is decreased with this PR, due to remapping now coverage is completely recalculated with ES6 output. When it comes to original typescript source, I could observe coverage is [around ~85%.](https://coveralls.io/builds/3869129)

**Comment by blesh** [comment]
This looks great so far. Unsure when I'll get a chance to merge this though. Might be a few days.

**Comment by kwonoj** [comment]
Take your time, PR'll be there when you're back :) If I could find solutions of converting sourcemaps, I'll update PR meanwhile

File: .travis.yml (hunk 4 above)
Comment (staltz) [inline .travis.yml]: :+1:

**Comment by blesh** [comment]
Also we'll have to research getting coveralls to be more lax until we're out of alpha

**Comment by kwonoj** [comment]
@blesh I could find this in coverall repo setting, maybe this is way to control? (Seems only repo owner can change those values)

image

**Comment by kwonoj** [comment]
PR's updated including one additional commit to close #558, now `cover` generates multi stage sourcemap pointing to original source. As same as original commit, it is expected to see coverage goes down compare to coverage against commonjs based source codes. You may able to check coverage report at https://coveralls.io/builds/3890453 (coverage check in above)

I was originally tried to suggest to move into different build system other than simple npm scripts, but found out that changes will include amount of out-of-scope changes not related with this specific issue. Those can be discussed in further as long term goal, focused to resolve immediate issue first.

**Comment by blesh** [comment]
Okay, I've temporarily increased thresholds of Coveralls to allow this to merge without breaking Travis. We'll see if it works.

**Comment by blesh** [comment]
@kwonoj so I pulled down the this PR, rebased on master, removed node_modules, reinstalled via npm, and ran `npm run build_cover` and it runs fine right to the end, and I get an error when it calls:

```
node lib/stage_sourcemap.js && remap-istanbul -i coverage/coverage.json -o coverage/coverage-remapped.json && remap-istanbul -i coverage/coverage.json -o coverage/coverage-remapped.lcov -t lcovonly && remap-istanbul -i coverage/coverage.json -o coverage/coverage-remapped -t html
```

Error: 

```
Error: Cannot find module 'jsonfile'
    at Function.Module._resolveFilename (module.js:336:15)
    at Function.Module._load (module.js:278:25)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/Users/blesh/GitHub/blesh/RxJS/lib/stage_sourcemap.js:3:16)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
```

I'll look into it, but not for too long, I have a LOT of PRs to catch up on. :)

**Comment by kwonoj** [comment]
Seems I mistakenly miss one of dependencies. I'll try to resolve on my side and update PR.

**Comment by blesh** [comment]
@kwonoj great! Notify me when you've got it. (Both here and Twitter today, because my GH inbox is dead from last week).

File: .travis.yml [hunk 5]
```diff
@@ -20,4 +20,4 @@ script:
   - npm run build_cover
 
 after_script:
-  - cat ./coverage/lcov.info | ./node_modules/coveralls/bin/coveralls.js
+  - cat ./coverage/coverage-remapped.lcov  | ./node_modules/coveralls/bin/coveralls.js
```
Comment (blesh) [inline .travis.yml]: Travis doesn't seem to be running the check now... this line is suspect.

Comment (kwonoj) [inline .travis.yml]: Yes, I'm looking those part now. On my forked branch this works, failed here, continuing investigation.

**Comment by kwonoj** [comment]
I will push some dummy commits to analyze logs, will flatten them once issue is resolved. Will notify you @blesh separately once it's done.

File: .travis.yml (hunk 5 above)
Comment (blesh) [inline .travis.yml]: Also, it seems that `remap-istanbul` isn't writing a `coverage-remapped.lcov` file at all.

**Comment by kwonoj** [comment]
OK, travis & coveralls both happy. Reason of failures was `jsonfile` module has unexpected regression on latest 2.2.3 version, pinned down into 2.2.2 as known good. Dummy commits are flattened.

**Comment by blesh** [comment]
merged with 69aa51d ... thanks for all of that @kwonoj! That was a tricky bit of work you did there.

**Pull request closed without merging**